sdk/schemaregistry/azure-schemaregistry/azure/schemaregistry/_encoder_protocols.py

```diff
@@ -0,0 +1,247 @@
+# --------------------------------------------------------------------------
+# Copyright (c) Microsoft Corporation. All rights reserved.
+# Licensed under the MIT License. See License.txt in the project root for
+# license information.
+# --------------------------------------------------------------------------
+from typing import Any, Dict, Mapping, Union, overload, Optional, Type
+from typing_extensions import Protocol, TypedDict  # type: ignore
+
+
+class SchemaContentValidate(Protocol):
+    def __call__(self, schema: Mapping[str, Any], content: Mapping[str, Any]) -> None:
+        """
+        Validates content against provided schema. If invalid, raises Exception.
+         Else, returns None.
+
+        :param mapping[str, any] schema: The schema to validate against.
+        :param mapping[str, any] content: The content to validate.
+
+        :rtype: None
+        :returns: None if valid.
+        :raises: Exception if content is invalid against provided schema.
+        """
+
+
+class MessageContent(TypedDict):
+    """A dict with required keys:
+    - `content`: bytes
+    - `content_type`: str
+    """
+
+    content: bytes
+    content_type: str
+
+
+class MessageType(Protocol):
+    """Message Types that set and get content and content type values internally."""
+
+    @classmethod
+    def from_message_content(
+        cls, content: bytes, content_type: str, **kwargs: Any
+    ) -> "MessageType":
+        """Creates an object that is a subtype of MessageType, given content type and
+         a content value to be set on the object.
+
+        :param bytes content: The content value to be set as the body of the message.
+        :param str content_type: The content type to be set on the message.
+        :rtype: ~azure.schemaregistry.encoder.jsonencoder.MessageType
+        :returns: The MessageType object with encoded content and content type.
+        """
+
+    def __message_content__(self) -> MessageContent:
+        """A MessageContent object, with `content` and `content_type` set to
+         the values of their respective properties on the MessageType object.
+
+        :rtype: ~azure.schemaregistry.encoder.jsonencoder.MessageContent
+        :returns: TypedDict of the content and content type from the MessageType object.
+        """
+
+
+class SchemaEncoder(Protocol):
+    """
+    Provides the ability to encode and decode content according to a provided schema or schema ID
+     corresponding to a schema in a Schema Registry group.
+    """
+
+    @overload
+    def encode(
+        self,
+        content: Mapping[str, Any],
+        *,
+        schema: str,
+        schema_id: None = None,
```

Comment:
is the typing here proper for  `schema_id`?

Reply:
yes, so this overload is when schema string is passed in, schema id is required to be None in that case.